File: download_problems.py
#!/usr/bin/env python3
"""
Script to download all Project Euler problems information.
Creates a problems.md file with titles and descriptions from problems 1 to 969.

Note: This script requires internet access to projecteuler.net
"""

import sys
import time

try:
    import requests
    from lxml import html as lxml_html
except ImportError:
    print("Error: Required packages not installed.", file=sys.stderr)
    print("Please install: pip install requests lxml", file=sys.stderr)
    sys.exit(1)


def get_problem_title(problem_ref):
    """
    Get the problem title from the main problem page.
    
    Args:
        problem_ref: The problem reference number (1-969)
    
    Returns:
        The problem title string, or None if not found
    """
    url = f"https://projecteuler.net/problem={problem_ref}"
    try:
        response = requests.get(url, timeout=30)
        response.raise_for_status()
        
        tree = lxml_html.fromstring(response.content)
        # Extract title from h2 tag at /html/body/div[1]/div/h2
        title_elements = tree.xpath('/html/body/div[1]/div/h2')
        
        if title_elements:
            title = title_elements[0].text_content().strip()
            return title
        return None
    except Exception as e:
        print(f"Error fetching title for problem {problem_ref}: {e}", file=sys.stderr)
        return None


def get_problem_description(problem_ref):
    """
    Get the problem description from the minimal page.
    
    Args:
        problem_ref: The problem reference number (1-969)
    
    Returns:
        The problem description as raw HTML, or None if not found
    """
    url = f"https://projecteuler.net/minimal={problem_ref}"
    try:
        response = requests.get(url, timeout=30)
        response.raise_for_status()
        
        tree = lxml_html.fromstring(response.content)
        # Extract description from /html/body/pre
        pre_elements = tree.xpath('/html/body/pre')
        
        if pre_elements:
            # Get the raw HTML content from the pre tag
            description_html = lxml_html.tostring(pre_elements[0], encoding='unicode', method='html')
            # Remove the <pre> and </pre> tags
            description_html = description_html.replace('<pre>', '').replace('</pre>', '').strip()
            return description_html
        return None
    except Exception as e:
        print(f"Error fetching description for problem {problem_ref}: {e}", file=sys.stderr)
        return None


def download_all_problems(start=1, end=969, output_file="problems.md"):
    """
    Download all problems and create a markdown file.
    
    Args:
        start: Starting problem number (default: 1)
        end: Ending problem number (default: 969)
        output_file: Output markdown file name (default: problems.md)
    """
    print(f"Downloading problems {start} to {end}...")
    print(f"Output file: {output_file}")
    
    with open(output_file, 'w', encoding='utf-8') as f:
        f.write("# Project Euler Problems\n\n")
        f.write(f"Problems {start} to {end}\n\n")
        f.write("---\n\n")
        
        for ref in range(start, end + 1):
            print(f"Processing problem {ref}...", end=' ')
            
            title = get_problem_title(ref)
            description = get_problem_description(ref)
            
            if title and description:
                f.write(f"## Problem {ref}: {title}\n\n")
                f.write(f"{description}\n\n")
                f.write("---\n\n")
                print("✓")
            else:
                print(f"✗ (title: {'✓' if title else '✗'}, description: {'✓' if description else '✗'})")
                if title:
                    f.write(f"## Problem {ref}: {title}\n\n")
                    f.write("*Description not available*\n\n")
                    f.write("---\n\n")
            
            # Be nice to the server - add a small delay
            time.sleep(0.5)
    
    print(f"\n✓ Done! Problems saved to {output_file}")


if __name__ == "__main__":
    # Allow command line arguments for start and end
    import argparse
    
    parser = argparse.ArgumentParser(description='Download Project Euler problems')
    parser.add_argument('--start', type=int, default=1, help='Starting problem number (default: 1)')
    parser.add_argument('--end', type=int, default=969, help='Ending problem number (default: 969)')
    parser.add_argument('--output', type=str, default='problems.md', help='Output file (default: problems.md)')
    parser.add_argument('--test', action='store_true', help='Test mode - only download first 5 problems')
    
    args = parser.parse_args()
    
    if args.test:
        print("Running in test mode - downloading only problems 1-5")
        download_all_problems(start=1, end=5, output_file=args.output)
    else:
        download_all_problems(start=args.start, end=args.end, output_file=args.output)

File: DownloadProblems/download_problems.cs
/*
 * Script to download all Project Euler problems information.
 * Creates a problems.md file with titles and descriptions from problems 1 to 969.
 * 
 * Note: This script requires internet access to projecteuler.net
 * 
 * Usage: 
 *   cd DownloadProblems && dotnet run -- [arguments]
 * 
 * Arguments:
 *   --start N       Starting problem number (default: 1)
 *   --end N         Ending problem number (default: 969)
 *   --output FILE   Output file (default: problems.md)
 *   --delay N       Delay between requests in seconds (default: 0.5)
 *   --test          Test mode - only download first 5 problems
 */

using System;
using System.Net.Http;
using System.Threading.Tasks;
using System.IO;

// Parse command line arguments
var commandArgs = Environment.GetCommandLineArgs().Skip(1).ToArray();
int start = 1;
int end = 969;
string outputFile = "problems.md";
double delay = 0.5;
bool testMode = false;

for (int i = 0; i < commandArgs.Length; i++)
{
    switch (commandArgs[i])
    {
        case "--start":
            if (i + 1 < commandArgs.Length)
            {
                if (!int.TryParse(commandArgs[++i], out start))
                {
                    Console.Error.WriteLine($"Error: Invalid value for --start: {commandArgs[i]}");
                    return;
                }
            }
            break;
        case "--end":
            if (i + 1 < commandArgs.Length)
            {
                if (!int.TryParse(commandArgs[++i], out end))
                {
                    Console.Error.WriteLine($"Error: Invalid value for --end: {commandArgs[i]}");
                    return;
                }
            }
            break;
        case "--output":
            if (i + 1 < commandArgs.Length) outputFile = commandArgs[++i];
            break;
        case "--delay":
            if (i + 1 < commandArgs.Length)
            {
                if (!double.TryParse(commandArgs[++i], out delay))
                {
                    Console.Error.WriteLine($"Error: Invalid value for --delay: {commandArgs[i]}");
                    return;
                }
            }
            break;
        case "--test":
            testMode = true;
            break;
        case "--help":
        case "-h":
            Console.WriteLine("Usage: cd DownloadProblems && dotnet run -- [arguments]");
            Console.WriteLine();
            Console.WriteLine("Arguments:");
            Console.WriteLine("  --start N       Starting problem number (default: 1)");
            Console.WriteLine("  --end N         Ending problem number (default: 969)");
            Console.WriteLine("  --output FILE   Output file (default: problems.md)");
            Console.WriteLine("  --delay N       Delay between requests in seconds (default: 0.5)");
            Console.WriteLine("  --test          Test mode - only download first 5 problems");
            Console.WriteLine("  --help, -h      Show this help message");
            return;
    }
}

if (testMode)
{
    Console.WriteLine("Running in test mode - downloading only problems 1-5");
    start = 1;
    end = 5;
}

await DownloadAllProblems(start, end, outputFile, delay);

static async Task<string?> GetProblemTitle(int problemRef, HttpClient client)
{
    string url = $"https://projecteuler.net/problem={problemRef}";
    try
    {
        var response = await client.GetAsync(url);
        response.EnsureSuccessStatusCode();
        
        var html = await response.Content.ReadAsStringAsync();
        
        // Simple parsing to extract title from h2 tag
        // Looking for pattern: <h2>Problem Title</h2>
        var h2Start = html.IndexOf("<h2>");
        if (h2Start >= 0)
        {
            h2Start += 4; // Move past <h2>
            var h2End = html.IndexOf("</h2>", h2Start);
            if (h2End > h2Start)
            {
                var title = html.Substring(h2Start, h2End - h2Start).Trim();
                // Decode HTML entities
                title = System.Net.WebUtility.HtmlDecode(title);
                return title;
            }
        }
        return null;
    }
    catch (Exception ex)
    {
        Console.Error.WriteLine($"Error fetching title for problem {problemRef}: {ex.Message}");
        return null;
    }
}

static async Task<string?> GetProblemDescription(int problemRef, HttpClient client)
{
    string url = $"https://projecteuler.net/minimal={problemRef}";
    try
    {
        var response = await client.GetAsync(url);
        response.EnsureSuccessStatusCode();
        
        var html = await response.Content.ReadAsStringAsync();
        
       
        return html;
    }
    catch (Exception ex)
    {
        Console.Error.WriteLine($"Error fetching description for problem {problemRef}: {ex.Message}");
        return null;
    }
}

static async Task DownloadAllProblems(int start, int end, string outputFile, double delay)
{
    Console.WriteLine($"Downloading problems {start} to {end}...");
    Console.WriteLine($"Output file: {outputFile}");
    
    using var client = new HttpClient();
    client.Timeout = TimeSpan.FromSeconds(30);
    
    using var writer = new StreamWriter(outputFile, false, System.Text.Encoding.UTF8);
    await writer.WriteLineAsync("# Project Euler Problems");
    await writer.WriteLineAsync();
    await writer.WriteLineAsync($"Problems {start} to {end}");
    await writer.WriteLineAsync();
    await writer.WriteLineAsync("---");
    await writer.WriteLineAsync();
    
    for (int problemRef = start; problemRef <= end; problemRef++)
    {
        Console.Write($"Processing problem {problemRef}... ");
        
        var title = await GetProblemTitle(problemRef, client);
        var description = await GetProblemDescription(problemRef, client);
        
        if (title != null && description != null)
        {
            await writer.WriteLineAsync($"## Problem {problemRef}: {title}");
            await writer.WriteLineAsync();
            await writer.WriteLineAsync(description);
            await writer.WriteLineAsync();
            await writer.WriteLineAsync("---");
            await writer.WriteLineAsync();
            Console.WriteLine("✓");
        }
        else
        {
            Console.WriteLine($"✗ (title: {(title != null ? "✓" : "✗")}, description: {(description != null ? "✓" : "✗")})");
            if (title != null)
            {
                await writer.WriteLineAsync($"## Problem {problemRef}: {title}");
                await writer.WriteLineAsync();
                await writer.WriteLineAsync("*Description not available*");
                await writer.WriteLineAsync();
                await writer.WriteLineAsync("---");
                await writer.WriteLineAsync();
            }
        }
        
        // Be nice to the server - add a small delay
        await Task.Delay(TimeSpan.FromSeconds(delay));
    }
    
    Console.WriteLine($"\n✓ Done! Problems saved to {outputFile}");
}
